File: cli/src/main/kotlin/com/maestrorecorder/cli/merger/ElementSelector.kt
package com.maestrorecorder.cli.merger

import com.maestrorecorder.shared.models.TapOnSelector
import com.maestrorecorder.shared.models.UiElement

object ElementSelector {

    fun selectBest(element: UiElement): TapOnSelector {
        val resourceId = element.resourceId
        val text = element.text
        val contentDescription = element.contentDescription

        // Priority 1: resource-id (strip package prefix for cleaner YAML)
        if (!resourceId.isNullOrBlank()) {
            val idSuffix = resourceId.substringAfter(":id/", "")
            val shortId = if (idSuffix.isNotEmpty()) idSuffix else resourceId
            return TapOnSelector.ById(shortId)
        }

        // Priority 2: visible text
        if (!text.isNullOrBlank()) {
            return TapOnSelector.ByText(text)
        }

        // Priority 3: content description
        if (!contentDescription.isNullOrBlank()) {
            return TapOnSelector.ByContentDescription(contentDescription)
        }

        // Priority 4: coordinate fallback
        return TapOnSelector.ByPoint(element.bounds.centerX, element.bounds.centerY)
    }
}

File: packages/server/src/recording/element-selector.ts
import type { TapOnSelector, UiElement } from '@maestro-recorder/shared';

export function selectBestSelector(element: UiElement | null, x: number, y: number): TapOnSelector {
  if (!element) return { kind: 'point', x, y };

  // Priority 1: resource-id (strip package prefix)
  if (element.resourceId) {
    const id = element.resourceId.includes(':id/')
      ? element.resourceId.split(':id/')[1]
      : element.resourceId;
    return { kind: 'id', id };
  }

  // Priority 2: visible text (non-empty, reasonable length)
  if (element.text && element.text.trim().length > 0 && element.text.length < 50) {
    return { kind: 'text', text: element.text };
  }

  // Priority 3: content description
  if (element.contentDescription) {
    return { kind: 'contentDescription', description: element.contentDescription };
  }

  // Priority 4: coordinate fallback
  return { kind: 'point', x, y };
}
